### ISSUE TITLE
Need unit tests for calc...() popgen functions

### ISSUE THREAD

**Issue by bhaller** [MesserLab/SLiM#474]
The need for unit tests for SLiM's popgen functions has been underlined by another discovery of a bug with them (https://groups.google.com/g/slim-discuss/c/Yacfk9EIYeU/m/bc72wVUzBAAJ).  I'm not sure how to test them, though.  I suppose a test could construct a population with known mutations, placed into the genomes at known positions/frequencies, and then test that the value calculated by the function matches the expected value calculated independently from first principles or by other software.  If someone can supply me with a test scenario and an expected value, I can construct a corresponding SLiM test, but I don't have the knowledge necessary to come up with appropriate scenarios and expected values.  These test scenarios wouldn't need to be large/complex; even a test with a genome of say, ten base positions long with, say, five mutations present and four diploid individuals (eight genomes) would be quite sufficient to test that the math and logic are correct, I would think.  It would be good to have such tests for all of the calc...() functions.  Perhaps @npb596 or @petrelharp or @philippmesser could help me with this?

**Comment by bhaller** [comment]
This could take the form of a VCF file and an expected value.  My SLiM test could simply load the VCF and check for a match (within reasonable numerical tolerance) to the expected value.

**Comment by bhaller** [comment]
OK.  Why not expected values from theory?

**Comment by petrelharp** [comment]
Because that is so much more complicated - you have to worry about statistical power; how close is "close enough"; etcetera. That sort of thing is good for validation, but not so good for unit tests (for one thing you end up having to run a lot of simulatiosn to make sure).  What we do in tskit, for instance, is usually just pull up the definition of the thing, then code up some real simple implementation that doesn't worry about efficiency; and compare to that. msprime does have a whole `validation.py` script that does statistical comparisons to other simulation software; but that's a much messier thing.

**Comment by bhaller** [comment]
> Because that is so much more complicated - you have to worry about statistical power; how close is "close enough"; etcetera. That sort of thing is good for validation, but not so good for unit tests (for one thing you end up having to run a lot of simulatiosn to make sure). What we do in tskit, for instance, is usually just pull up the definition of the thing, then code up some real simple implementation that doesn't worry about efficiency; and compare to that. msprime does have a whole `validation.py` script that does statistical comparisons to other simulation software; but that's a much messier thing.

Aha, I see.  Yes, there are certainly problems with doing statistical tests for validation.  SLiM already does tons of them, though.  But if a precise comparison to the "right answer" is possible, that's certainly better!

**Comment by npb596** [comment]
Sorry for not responding to this until now. I saw it at the time but knew I'd be rather busy for the rest of that term, and I was, so I left it alone. I agree that comparison to other calculated values makes sense and sounds relatively doable. I don't know what computing expected values from theory would entail personally. I'm responding now because I did write a script with a simple sort of test. I did tests like it when writing the functions but never had a clean version to show until now. I'm attaching the script as a text file so Github lets me but it should really have a .slim suffix.

[Test_PopGen_Functions.txt](https://github.com/user-attachments/files/18413581/Test_PopGen_Functions.txt)

It will run a simple neutral simulation with either a population contraction or expansion (one should be commented out at a time) and plot Tajima's D against time (in ticks). The basic qualitative expectation is that a contraction should increase Tajima's D and an expansion should decrease it. The change in pop size occurs halfway along the x-axis in the plot and it does seem to exhibit this pattern. Obviously there's also random variation in Tajima's D so sometimes the change halfway isn't obvious. I assume increasing pop size would reduce the variation but the simulation takes about a minute as is so I didn't want to push parameters too much. There's comments with specific parameters in script. Also, I ran it on the current release build so I define calcTajimasD2 and calcPi2 explicitly in there, but I took them straight from [slim_functions.cpp](https://github.com/MesserLab/SLiM/blob/9567f765518512bb924ec508c8074604d5916b4f/core/slim_functions.cpp#L362) so should match what's supposed to be built-in barring the current bug you noted.

The above could also be used to plot calcPi and calcWattersonsTheta. The changes in Tajima's D should correspond to changes in these. Also, in neutrality with constant pop size, both of those should fluctuate around $4Nm$ where N would be pop size and m is mutation rate. Tajima's D should fluctuate around 0. Checking those was something else I did in early testing, where it seemed good, but should be checked again. I guess you could incorporate selection to see if Tajima's D matches expectations there but that gets a bit more complicated.

Regarding comparisons to inference programs; SLiM can output VCFs so I assume those can be analyzed directly with something like VCFtools, which computes pi and Tajima's D. So someone could just compute the stats within SLiM at the time a VCF is output, and run VCFtools (or other programs) on that same VCF and see if they are about the same, right? For a different project I did some comparisons between software and used a dummy VCF I made matching a textbook example. I believe it was an alignment given in figure 3.1 of Matt Hahn's Molecular Population Genetics. I don't have the book on hand. The textbook does the pencil-and-paper pi and watterson's theta calcs on that alignment so I assume that would work as something with known mutations in known positions/frequencies. I could dig around for the dummy VCF.

Some of that I could probably do soon if it'd be of interest. I'd be happy to let someone else do it too but I'm assuming no one's biting if the post has been up this long.

**Comment by bhaller** [comment]
@npb596 I'm aiming to release SLiM 5 at the beginning of April.  It'd be great to have these unit tests in there, so we're more confident that the calc...() functions are working as expected.  Have you had a chance to work on this at all?

**Comment by npb596** [comment]
@bhaller sorry about the delay, I've been prepping for a conference I'm going to tomorrow and which will last for the whole week. I can finalize the changes you made to the test scripts next week for certain and I do have some work towards testing a simple VCF that I should also be able to finish up before the end of next week.

**Comment by bhaller** [comment]
@npb596 Have fun at ProbGen, I assume that's where you are!  :->

**Comment by bhaller** [comment]
Hi @npb596!  Have a good ProbGen?  Just a ping on this, since I'm working on wrapping up SLiM 5.  :->

**Comment by npb596** [comment]
Yep, it was great, had the weekend in NY city and just got back home yesterday! I'm just about ready to add a new pull request to SLiM-Extras, so details to come there soon.

**Comment by bhaller** [comment]
This issue was fixed by @npb596 in SLiM-Extras, in PR https://github.com/MesserLab/SLiM-Extras/pull/34.
